### PR TITLE
JRoute between clients (site -> admin | admin -> site)

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Improvement.
#### Summary of Changes

This PR is for allowing JRoute to generate URL's from admin to site and vice-versa.

This is a current problem especially if you are trying to make links from the backend to the frontend.

So, with this PR, there is no need anymore to make special code just to get the frontend link from the backend. With this PR we just need to use JRoute, like all other links.

The same will work to links from frontend to backend.
After this PR, no need to hardcode frontend to admin links.
##### Test 1: Links from admin to site
- Use latest staging with sample data from at least two languages (en-GB, fr-FR)
- Add to isis index.php file...

``` php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Admin to Admin: ' . JRoute::_('index.php'), 'notice');
$app->enqueueMessage('2. Admin to Admin: ' . JRoute::_('index.php?option=com_content'), 'notice');
$app->enqueueMessage('3. Admin to Site: ' . JRoute::_('index.php?option=com_content', true, null, 'site'), 'notice');
$app->enqueueMessage('4. Admin to Site (MutilLanguage 1): ' . JRoute::_('index.php?option=com_content&lang=fr', true, null, 'site'), 'notice');
$app->enqueueMessage('5. Admin to Site (MutilLanguage 2): ' . JRoute::_('index.php?option=com_users&lang=en', true, null, 'site'), 'notice');
```
- Open any admin page. Check the URL in resulting notice message are correct.

Note: You should test also other URLs (articles, menu items, etc).

Note 2: Try with the several variation of SEF/non-SEF,  with or without URL rewritting,  with or without multilanguage (i.e., with or without language filter plugin enabled).
##### Test 2: Links from site to admin

Add to protostar index.php file...

``` php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Site to Site: ' . JRoute::_('index.php'), 'notice');
$app->enqueueMessage('2. Site to Site: ' . JRoute::_('index.php?option=com_content'), 'notice');
$app->enqueueMessage('3. Site to Admin: ' . JRoute::_('index.php?option=com_content', true, null, 'administrator'), 'notice');
```
- Open any frontend page. Check the URL in resulting notice message are correct.

Note: You should test also other URLs.

Note 2: Try with the several variation of SEF/non-SEF,  with or without URL rewritting,  with or without multilanguage (i.e., with or without language filter plugin enabled).
##### Test 3: Code review

Do a code review.
##### Test 4: Current URLs

Check if current URL generated by JRouter are exactly the same, i.e., no change in current URL.
Probably the best is to leave the patch applied for some time, while you do other things and check there are no problems with the current URLs.
#### Observations

Thanks to @mbabker, @ggppdk and @infograf768 for helping out with this one.

Mantainers, to make sure that are no unintended consequences, it would be good if this can be also reviewed from someone that understand the Router better than i do.

### CODE REVIEW

**Comment by andrepereiradasilva** [comment]
travis error: 

```
PHP Fatal error:  Cannot instantiate abstract class JSessionStorage in /home/travis/build/joomla/joomla-cms/libraries/joomla/session/storage.php on line 85
```

? nothing to do with this PR i think

**Comment by mahagr** [comment]
I did a quick code review and I really like the implementation. Thanks for implementing this!

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 4502d6f20a0fb3389d69fabf6da45b6c556a0719<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by andrepereiradasilva** [comment]
can anyone force a travis rebuild here?

**Comment by brianteeman** [comment]
Restarted travis as requested

**Comment by andrepereiradasilva** [comment]
thansk @brianteeman 

same failure ...

`PHP Fatal error:  Cannot instantiate abstract class JSessionStorage in /home/travis/build/joomla/joomla-cms/libraries/joomla/session/storage.php on line 85`

i don't understand with this happens.
can anyone help with this?

**Comment by brianteeman** [comment]
First thought is that your branch is not upto date (thats been the problem
in the past)

On 10 July 2016 at 10:49, andrepereiradasilva <notifications@github.com>
wrote:

> thansk @brianteeman <https://github.com/brianteeman>
>
> same failure ...
>
> PHP Fatal error: Cannot instantiate abstract class JSessionStorage in
> /home/travis/build/joomla/joomla-cms/libraries/joomla/session/storage.php
> on line 85
>
> i don't understand with this happens.
> can anyone help with this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11060#issuecomment-231580295>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPH8T9Pvky4UuIKsbnKeQOUlDCM97zAks5qUMAUgaJpZM4JIl1C>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by andrepereiradasilva** [comment]
well but it is up to date, so is not that.

looking at the error call stack https://travis-ci.org/joomla/joomla-cms/jobs/143540387#L962 it seems to be related with JRoute in opensearch ... ?

**Comment by brianteeman** [comment]
Now you are beyond where I can help - sorry

Maybe @mbabker or @wilsonge have an idea

On 10 July 2016 at 10:53, andrepereiradasilva <notifications@github.com>
wrote:

> well but it is up to date, so is not that.
>
> looking at the error call stack
> https://travis-ci.org/joomla/joomla-cms/jobs/143540387#L962 it seems to
> be related with JRoute in opensearch ... ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11060#issuecomment-231580449>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPH8R636lcRvXVyq_nvKY9X6niYHJNLks5qUMEYgaJpZM4JIl1C>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by yvesh** [comment]
@andrepereiradasilva The thing is by default with `JApplicationCms::getInstance($client)` you create a new session (except for cli).

Somehow the session handler (fallback 'none') became an empty String '' after you changes. 

$handler = $conf->get('session_handler', 'none');

Which causes that it tries to instantiate the abstract class JSessionStorage. (There is no isAbstract() / isInstantiable() test before).

**Comment by andrepereiradasilva** [comment]
@yvesh thanks

but i didn't change any session related stuff as you can see from the code changes https://github.com/joomla/joomla-cms/pull/11060/files

**Comment by andrepereiradasilva** [comment]
seems fine now. thanks again for the hints @yvesh .

**Comment by wilsonge** [comment]
@hackwar can you please review?

**Comment by Hackwar** [comment]
On a first quick look across the code, it looks pretty good, however I see one issue with backwards compatibility. Changing JRoute::$_router from an object to an array has in the past been defined as a breach in backwards compatibility.

I fear a few issues with problematic component routers, but I will have to take a deeper look into this first.

**Comment by mahagr** [comment]
Huh.. How can private variable cause backwards compatibility break as it cannot be accessed from anywhere else but this class? It cannot be used even from child classes and therefore it cannot break anything.

**Comment by Hackwar** [comment]
Sorry, you are right. I missed that it was private and not just protected.

**Comment by infograf768** [comment]
The correct tests for multilanguage to add in index.php should be
```
$app->enqueueMessage('4. Admin to Site (MutilLanguage 1): ' . JRoute::_('index.php?option=com_content&lang=fr-FR', true, null, 'site'), 'notice');
$app->enqueueMessage('5. Admin to Site (MutilLanguage 2): ' . JRoute::_('index.php?option=com_users&lang=en-GB', true, null, 'site'), 'notice');
```
The reason is that someone may use en-US and en-GB, or fr-FR and fr-CA, etc.

Test successful here.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on d07ee167eb3c42a8c9dfd276fd7e53bbb47be886<br /><br />Great!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by infograf768** [comment]
See example of use here:
https://github.com/joomla/joomla-cms/pull/10932#issuecomment-231678977

**Comment by hardiktailored** [comment]
I have tested and it seems it works for front-end with SEF and URL Rewriting but for back-end URL(ADMIN to SITE) remains same. Is that correct?

See front-end:
screen shot 2016-07-13 at 05 52 05

See Back-end:
screen shot 2016-07-13 at 05 52 21<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by andrepereiradasilva** [comment]
> I have tested and it seems it works for front-end with SEF and URL Rewriting but for back-end URL(ADMIN to SITE) remains same. Is that correct?

@hardiktailored no is not correct. you should also have SEF URL from admin to site

can you tell which options did you use for global config SEF? if you are using multilanguage and if you have the language filter plugin enabled.

**Comment by Hackwar** [comment]
administrator has no SEF URLs.

**Comment by andrepereiradasilva** [comment]
@Hackwar i think he's talking about admin to site URLs.

**Comment by hardiktailored** [comment]
@andrepereiradasilva Yes, exactly. I think SITE URLs should also be rewrite when SEF and Rewriting enabled even when linking from ADMIN.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by andrepereiradasilva** [comment]
and it should be, that's why i asked you those questions:
> can you tell which options did you use for global config SEF? if you are using multilanguage and if you have the language filter plugin enabled.

**Comment by hardiktailored** [comment]
I have language filter plugin disabled so ignore 4th and 5th URL but don't you think 3rd URL should be changed?
screen shot 2016-07-13 at 07 48 57<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by andrepereiradasilva** [comment]
with that config, with or without language filter plugin enabled, all 3 should be changed.
strange i can't reproduce your issue
are you sure you have the patch applied?

**Comment by hardiktailored** [comment]
Enabling Language Filter plugin, URL rewrites. But that should also work without need of this plugin enabled.
screen shot 2016-07-13 at 08 02 04<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by hardiktailored** [comment]
Yes, patch applied.
screen shot 2016-07-13 at 08 04 53<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11060">issues.joomla.org/joomla-cms/11060</a>.</sub>

**Comment by andrepereiradasilva** [comment]
ok. will check again

**Comment by andrepereiradasilva** [comment]
you're right, somehow it doesn't work now without the language filter plugin enabled :confused:

also i discovered another issue when you have an installed language that does not have a content language we get a PHP notices

```
Notice: Undefined index: pt-BR in /path/to/joomla-staging/plugins/system/languagefilter/languagefilter.php on line 190
Notice: Trying to get property of non-object in /path/to/joomla-staging/plugins/system/languagefilter/languagefilter.php on line 190
Notice: Undefined index: pt-BR in /path/to/joomla-staging/plugins/system/languagefilter/languagefilter.php on line 190
Notice: Trying to get property of non-object in /path/to/joomla-staging/plugins/system/languagefilter/languagefilter.php on line 190
```

so clearly more improvements are needed

**Comment by andrepereiradasilva** [comment]
hum if i revert this change it works again with the language filter plugin disabled.
https://github.com/joomla/joomla-cms/pull/11060/commits/d07ee167eb3c42a8c9dfd276fd7e53bbb47be886

Any thoughts @yvesh

File: libraries/cms/router/site.php [hunk 1]
```diff
@@ -162,8 +162,16 @@ public function build($url)
 			}
 		}
 
+		$basePath = JUri::base(true);
+
+		// If the site router is called from administrator client.
+		if (JPATH_BASE == JPATH_ADMINISTRATOR)
+		{
+			$basePath = str_replace('/administrator', '', $basePath);
```
Comment (izharaazmi) [inline libraries/cms/router/site.php]: On an installation where $basePath with /administrator anywhere in the path will be affected. See #8587 note.
Why don't we use `$basePath = JUri::root(true);` here?

**Comment by Fedik** [comment]
hm, it seems stopped to work, or I just confused 😄 

I tried test on sample data (in backend):
```php
JRoute::_('index.php?option=com_content&view=article&id=17', true, null, 'site')
```
and expected to see: `/article-category-blog/17-first-blog-post`
but got `/index.php?option=com_content&view=article&id=17&Itemid=435`, without SEF, and with wrong `Itemid`.

Maybe because some recent router changes?

**Comment by andrepereiradasilva** [comment]
hum .. @Hackwar can you help here?

**Comment by Hackwar** [comment]
Sorry, I don't know and unfortunately also don't have the time to investigate this.

**Comment by andrepereiradasilva** [comment]
Ok.thanks anyway

**Comment by mbabker** [comment]
This PR needs a new owner or it will have to be closed as abandoned.

**Comment by izharaazmi** [comment]
I'd like to take the charge and little rework, if acceptable?

**Comment by wilsonge** [comment]
I'll take this one up

**Comment by wilsonge** [comment]
Take it

**Comment by izharaazmi** [comment]
@wilsonge Sorry I didn't know you already took it. Email notification is not sent on self assign. I just responded to Michael. Please let me know if you'd like to proceed with this.

**Comment by wilsonge** [comment]
No it's absolutely fine :) Take it!

**Comment by andrepereiradasilva** [comment]
closing as @izharaazmi took over this
see https://github.com/joomla/joomla-cms/pull/16879

**Pull request closed without merging**